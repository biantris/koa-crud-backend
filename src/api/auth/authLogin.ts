import User from "../../modules/user/UserModel";
import { generateToken } from "../../auth/generateToken";

export const authLogin = async (ctx) => {
  const { email, password } = ctx.request.body;

  if (!email || !password) {
    ctx.status = 401;
    ctx.body = {
      message: "Email or password incorrect",
    };
    return;
  }

  const user = await User.findOne({
    email,
    removedAt: null,
  });

  if (!user) {
    ctx.status = 401;
    ctx.body = {
      message: "Email or password incorrect",
    };
    return;
  }

  let correctPassword;

  try {
    correctPassword = user.authenticate(password);
  } catch {
    ctx.status = 401;
    ctx.body = {
      message: "Email or password incorrect",
    };
    return;
  }

  if (!correctPassword) {
    ctx.status = 401;
    ctx.body = {
      message: "Email or password incorrect",
    };
    return;
  }

  ctx.status = 200;
  ctx.body = {
    message: "User authenticated with success",
    token: generateToken(user),
  };
};
