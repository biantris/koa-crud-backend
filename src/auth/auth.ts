import { Base64String, unbase64 } from "./base64";
import { getObjectId } from "../../test";
import User from "../modules/user/UserModel";

export type Token = {
  userId: string | null;
};

export const getToken = (authorization: Base64String): Token => {
  const userId = unbase64(authorization);

  if (!userId) {
    return {
      userId: null,
    };
  }

  return {
    userId,
  };
};

export const getUser = async (token: string) => {
  const { userId } = getToken(token);

  if (!userId) {
    return null;
  }

  const user = await User.findOne({
    _id: getObjectId(userId),
    removedAt: null,
  });

  return user;
};

export const auth = async (ctx, next) => {
  const { authorization } = ctx.header;

  if (!authorization) {
    ctx.status = 401;
    ctx.body = {
      message: "Unauthorized",
    };
    return;
  }

  const user = await getUser(authorization);

  if (!user) {
    ctx.status = 401;
    ctx.body = {
      message: "Unauthorized",
    };
    return;
  }

  ctx.user = user;

  await next();
};
