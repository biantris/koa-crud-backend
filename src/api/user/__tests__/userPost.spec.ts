import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  createApiCall,
  sanitizeTestObject,
} from "../../../../test";
import { createUser } from "../../../modules/user/fixtures/createUser";
import { base64 } from "../../../auth/base64";

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

const url = "/api/user";

it("should return error if authorization header does not exist", async () => {
  const response = await createApiCall({ url, payload: {} });

  expect(response.status).toBe(401);
  expect(response.body.message).toBe("Unauthorized");
  expect(response.body).toMatchSnapshot();
});

it("should return error if user it was not passed", async () => {
  const admin = await createUser({});
  const authorization = base64(`${admin._id}`);

  const response = await createApiCall({ url, payload: {}, authorization });

  expect(response.status).toBe(400);
  expect(response.body.message).toBe("User is required");
  expect(response.body).toMatchSnapshot();
});

it("should return error if user name it was not passed", async () => {
  const admin = await createUser({});
  const authorization = base64(`${admin._id}`);

  const user = { email: "test@test.com", password: "123456" };
  const response = await createApiCall({
    url,
    payload: { user },
    authorization,
  });

  expect(response.status).toBe(400);
  expect(response.body.message).toBe("name is a required field");
  expect(response.body).toMatchSnapshot();
});

it("should return error if user email it was not passed", async () => {
  const admin = await createUser({});
  const authorization = base64(`${admin._id}`);
  const user = { name: "obi wan kenobi", password: "123456" };
  const response = await createApiCall({
    url,
    payload: { user },
    authorization,
  });

  expect(response.status).toBe(400);
  expect(response.body.message).toBe("email is a required field");
  expect(response.body).toMatchSnapshot();
});

it("should return error if user password it was not passed", async () => {
  const admin = await createUser({});
  const authorization = base64(`${admin._id}`);

  const user = { name: "obi wan kenobi", email: "test@test.com" };
  const response = await createApiCall({
    url,
    payload: { user },
    authorization,
  });

  expect(response.status).toBe(400);
  expect(response.body.message).toBe("password is a required field");
  expect(response.body).toMatchSnapshot();
});

it("should create user with success", async () => {
  const admin = await createUser({});
  const authorization = base64(`${admin._id}`);
  const user = {
    name: "obi wan kenobi",
    email: "test@test.com",
    password: "123456",
  };

  const response = await createApiCall({
    url,
    payload: { user },
    authorization,
  });

  expect(response.status).toBe(200);
  expect(response.body.user.name).toBe(user.name);
  expect(response.body.user.email).toBe(user.email);
  expect(sanitizeTestObject(response.body)).toMatchSnapshot();
});

it("should return error is email is already in use", async () => {
  const admin = await createUser({});
  const authorization = base64(`${admin._id}`);
  await createUser({
    name: "obi wan kenobi",
    email: "test@test.com",
    password: "123456",
  });

  const userB = {
    name: "jay prichett",
    email: "test@test.com",
    password: "654321",
  };

  const response = await createApiCall({
    url,
    payload: { user: userB },
    authorization,
  });

  expect(response.status).toBe(400);
  expect(response.body.message).toBe("Email already in use");
  expect(sanitizeTestObject(response.body)).toMatchSnapshot();
});
