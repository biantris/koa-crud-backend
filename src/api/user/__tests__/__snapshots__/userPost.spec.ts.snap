// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create user with success 1`] = `
Object {
  "user": Object {
    "_id": "60af0d926fa981e9c07ac604",
    "email": "test@test.com",
    "name": "obi wan kenobi",
  },
}
`;

exports[`should return error if user email it was not passed 1`] = `
Object {
  "message": "email is a required field",
}
`;

exports[`should return error if user it was not passed 1`] = `
Object {
  "message": "User is required",
}
`;

exports[`should return error if user name it was not passed 1`] = `
Object {
  "message": "name is a required field",
}
`;

exports[`should return error if user password it was not passed 1`] = `
Object {
  "message": "password is a required field",
}
`;
